### DISCUSSION TITLE
Allow actions to set "neutral" conclusion

### DISCUSSION THREAD

**Discussion by jalaziz** [community/community#9875] [Actions]
Today, it seems there is no good or easy way to conclude a Github Action as anything other than a success or failure. However, the Checks API supports other conclusions that can be useful for an Action.

There have been [several][1] [discussions][2] [around][3] [this][4], and it seems that the ability to set the `neutral` conclusion by way of exit code was removed. The primary use case in some of the discussions seems to have been skipping a workflow. However, there are other uses for `neutral` that aren't solved by skipping.

For example, we would like to introduce an Action that confirms a commit message meets certain criteria. However, some of the criteria is optional, but we'd like to _warn_ developers when they're missing. In this case, failing the workflow is too harsh, and success doesn't seem correct either.

It would be great if a `neutral` or `warning` conclusion could be set through a workflow command or some other mechanism.

[1]: https://github.com/actions/toolkit/issues/146
[2]: https://github.community/t/github-actions-neutral-exit-code-is-incorrectly-interpreted-as-failure/16088/4
[3]: https://github.community/t/actions-neutral-exit-strategy/129800/4
[4]: https://github.com/actions/github-script/issues/147

**Comment by lhstrh** [comment]
Our project would benefit from having this feature as well. I arrived here via https://github.com/isaacs/github/issues/1465 which describes the same need.

**Comment by ethomson** [comment]
Thanks, @jalaziz - I agree that this would be useful.  I'm sorry that we haven't gotten to this yet (it actually touches a lot of different pieces in the platform) but this is on our backlog to improve in the future.  Thanks again.

**Reply by mlocati** [reply]
@ethomson Just a side note: there's the open issue https://github.com/actions/toolkit/issues/399 where a ton of users has been asking for something like this for a couple of years.

**Reply by pboling** [reply]
Worth noting that there are around 500 interactions on that actions/toolkit issue!  So many projects making commits referencing it...

When people make commits decrying the absurdity of this missing feature, please be sure to reference this feedback discussion also.

**Reply by eregon** [reply]
I created a separate discussion about this, which really focus on the heart of that issue: https://github.com/github/feedback/discussions/15452
This discussion seems rather unspecific, and I think showing exactly what needs to improve is more helpful.

**Comment by pboling** [comment]
The absence of this feature has wasted hours of my time while supporting community ruby gems, like `oauth` and `oauth2`, that have ostensibly more important things I could work on, than futzing with Github Actions.

I hope it makes it up the backlog quickly!

**Comment by carmanchris31** [comment]
EDIT: It looks like this is actually supported already?
> The final conclusion of the check. Can be one of action_required, cancelled, failure, neutral, success, skipped, stale, or timed_out
> —[source](https://docs.github.com/en/rest/checks/runs#:~:text=be%20one%20of-,action_required%2C%20cancelled%2C%20failure%2C%20neutral%2C%20success%2C%20skipped%2C%20stale,-%2C%20or%20timed_out.%20When)

**Reply by lucasff** [reply]
A check run isn't the same as a step on a GitHub Action workflow that is allowed to fail.
Microsoft, pretty please?

**Comment by pboling** [comment]
@eregon has created another Feedback Discussion that is more explicitly targeted to the ask for an `allow_failure`:
https://github.com/github/feedback/discussions/15452

**Reply by daira** [reply]
That is not the same thing. This issue is about explicitly posting a neutral conclusion for a job. `allow_failure` is a special case that would be trivial to implement in terms of the feature asked for here.

It's possible to post a neutral conclusion, but only at the expense of needing write permission on the target repo (usually achieved by having the job run on `pull_request_target`), which is an unacceptable security risk for many projects. For completeness and since it may be a sufficient workaround for some users, I'll show how to do that:
```
name: Demonstrate posting a three-way conclusion

# warning: do not just copy/paste this without understanding security risks
on:
  pull_request_target:

permissions:
  contents: read
  checks: write  # Required to post the status via API

jobs:
  the_job:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout repository
        uses: actions/checkout@v6.0.2 # pin this if needed

      - name: Name of the step
        id: the_step
        shell: bash
        run: |
          # Turn off immediate exit on error so we can capture the code
          set +e
          # exit code 0 = success, 1 = neutral, other = failure
          # script gets write access to the repo, even if it shouldn't need it :-(
          ./the_step.sh
          echo "exit_code=$?" >> "$GITHUB_OUTPUT"

      - name: Post status
        if: always()
        uses: actions/github-script@v7 # pin this if needed
        env:
          EXIT_CODE: ${{ steps.the_step.outputs.exit_code }}
        with:
          script: |
            const { owner, repo } = context.repo;
            const exitcode = parseInt(process.env.EXIT_CODE);

            let conclusion = 'failure';
            let title = 'Title to use for failure';

            if (exitcode === 0) {
              conclusion = 'success';
              title = 'Title to use for success';
            } else if (exitcode === 1) {
              conclusion = 'neutral';
              title = 'Title to use for neutral';
            }

            await github.rest.checks.create({
              owner,
              repo,
              name: 'Name of the step',
              head_sha: context.sha,
              status: 'completed',
              conclusion,
              output: {
                title: title,
                summary: `Script exited with code ${exitcode}.`
              }
            });
```

**Reply by pboling** [reply]
Agreed - not the same. It might help some who end up here though, since GitHub is unlikely to ever improve the platform. :(

**Comment by AleksaRistic216** [comment]
Four years later, 100+ upvotes, do we have this?
